e2e/regressions/build.go

```diff
@@ -56,6 +57,37 @@ func (c *regressionsTests) issue4203(t *testing.T) {
 	)
 }
 
+// This test will build a sandbox, as a non-root user from a dockerhub image
+// that contains a single folder and file with `000` permissions.
+// To verify we force files to be accessible / moveable / removable by the user
+// we check for `700` and `400` permissions on the folder and file respectively.
+func (c *regressionsTests) issue4517(t *testing.T) {
+	sandbox := filepath.Join(c.env.TestDir, "issue_4203")
+
+	c.env.RunSingularity(
+		t,
+		e2e.WithPrivileges(false),
+		e2e.WithCommand("build"),
+		e2e.WithArgs("--sandbox", sandbox, "docker://sylabsio/issue4517"),
```

Comment:
I don't want to derail this, but to the best of my understanding, the issue reported in #4517 can be reproduced using this definition file:

```
bootstrap: scratch

%setup
        mkdir ${SINGULARITY_ROOTFS}/d
        touch ${SINGULARITY_ROOTFS}/d/file
        chmod 000 ${SINGULARITY_ROOTFS}/d/file
        chmod a-w ${SINGULARITY_ROOTFS}/d
```

The resulting sandbox cannot be deleted:

```
$ ./builddir/singularity build --fakeroot --sandbox issue-4517.s ./issue4517.def
Build target already exists. Do you want to overwrite? [N/y] y
INFO:    Starting build...
INFO:    Running setup scriptlet
+ mkdir rootfs-120641e2-dff4-11e9-9ab5-54ee757bbcd2/d
+ touch rootfs-120641e2-dff4-11e9-9ab5-54ee757bbcd2/d/file
+ chmod 000 rootfs-120641e2-dff4-11e9-9ab5-54ee757bbcd2/d/file
+ chmod a-w rootfs-120641e2-dff4-11e9-9ab5-54ee757bbcd2/d
INFO:    Creating sandbox directory...
INFO:    Build complete: issue-4517.s

$ rm -rf issue-4517.s/
rm: cannot remove 'issue-4517.s/d/file': Permission denied
```

After fixing the permissions on the directory, the sandbox can be removed.

Is that what we are trying to fix / test in this PR or is that something different?

Reply:
To my mind that's something a bit different, as the user is explicitly executing `chmod 000` instructions in a file they are directly providing for the build, but it's worth considering in future.

In #4517 we have no `--fakeroot` and no def file. The user is able to get something that they cannot delete when they have not explicitly provided instructions to create `000` files / dirs - but simply built a local sandbox of a docker image.

Reply:
I'm not sure it's different.

My understanding of the issue is that the sandbox is created in a temporary directory, and if that directory is located in a different filesystem than the final destination, the move operation from the temporary location to the final location fails because it becomes a copy instead of a move. The reason for failing is because the original container image has a non-readable file.

Starting with the above, you can do:

```
$ singularity build --fakeroot issue-4517.sif ./issue-4517.def
...

$ singularity build --sandbox issue-4517.s issue-4517.sif
INFO:    Starting build...
INFO:    Creating sandbox directory...
FATAL:   While performing build: sandbox assemble failed: exit status 1: mv: cannot open '/tmp/sbuild-878592656/fs/d/file' for reading: Permission denied
```

that's the issue as reported, right?

Notice that master does not need this fix (it has already fixed the above issue in a different way). I'm a little concerned that with the proposed fix we would be doing things in a slightly different way in 3.4 and master (3.4 would be mucking around with permissions, master would not). And also about the fact that we are applying this fix based on the bootstrap value (the above steps are equivalent to `boostrap: localimage / from: my.sif`, correct?).

Also note that the test as proposed is not testing for this issue, as it's not making sure that the triggering condition exists (/tmp is on a different filesystem). The test as proposed is testing for the condition I wrote in the original comment (final permissions are "incorrect"), which you are saying is something different.

Reply:
We still have to mess with the permissions in master... otherwise the sandbox cannot be removed easily by the user as there could by `000` files there etc. master will still needs this fix from this PR, even if there is no `mv` of the sandbox, so it doesn't result in a build failure - so the intention is to do the same perm mangling in master.

The test is correct for verifying perms in master... and I'd argue that (albeit indirectly) it does ensure there wouldn't be a build failure if `/tmp` is on a different filesystem - with the current 3.4 code flow. If you feel very strongly about it looking directly into the `/tmp` rootfs on 3.4 then I can write a test for that - but it'll be more difficult, and that test is only useful on 3.4 - it won't work on master.

Regarding whether we do the permission mangling in the OCI extraction, or later (which would fix your def file example perms)... I'd still argue here that the user in the def file case is explicitly asking for a `chmod 000` in something they are directly providing to singularity - so it's reasonable to honor that. We *have* to do the perm changes on the OCI side though, otherwise we end up with straight builds from docker images being having perms preventing easy copy/move/delete, which isn't at all friendly for the user.

Reply:
Pardon my brain dump, I'm trying to narrow down what's bothering me.

Perhaps I'm confused regarding which issue we are trying to address.

There are two issues:

1. Failure to build because the move operation fails
2. Unable to remove the resulting sandbox directory easily

If we start from the .def file I provided, at some point during the build there's an intermediate location which contains a directory (d) and a file (d/file) with "funny" permissions (1. above). Somehow are are able to work with that, including reading the files to push them to the squashfs image (why? --fakeroot?) and removing the temporary directory when we are done (IIRC there's code in os.RemoveAll that's doing this, I mean trying to remove a directory, failing, and tweaking permissions so that it works... or it could be --fakeroot, too). In this case, we are happy to leave those funny permissions alone (2. above).

If instead of using the .def file as input, we use another image as input, we are saying we do have to tweak the permissions. I agree that the user is not being as explicit as in the first case, but the user is still saying "use that image as input", not "use that image as input and tweak it a little bit if necessary". That's the bit that's a little uncomfortable in my mind, as introducing this intermediate step is causing the resulting sandbox image to be different.

Thinking out loud: print a warning for every file where we are tweaking permissions, so that the user has a chance to know what happened?

Only 2. above affects master. In 3.4 (this PR) we are fixing both by tweaking the permissions.

And trying to go back to my original comment: this is completely optional, but I would rather avoid hitting the network unless it's necessary. I agree that a test for 1. above is complicated (mount tmpfs filesystem somewhere, point SINGULARITY_TEMPDIR there?) but since hitting the network is not changing this, I would _prefer_ to go with the version that starts from the .def file, creates a .sif and then creates a sandbox out of that .sif so that it can test that the resulting permissions in the sandbox have been correctly tweaked. That would be testing 2. above, right?

Anyhow, I'm not asking for any changes here, just trying to figure things out.

Reply:
Hi @mem - I am going to mark this as resolved from the other discussion that has taken place - to clean up the thread for people newly following through it. I've summarized this discussion below, please let me know if there's anything inaccurate, or you want to raise there.